# OS Walkthrough

In [3]:
!pip install opensearch-py

## Libraries Used

In [160]:
!pip install uuid

Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://semple_devs:****@nexus.global.lmco.com/repository/pypi-proxy/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for uuid: filename=uuid-1.30-py3-none-any.whl size=6501 sha256=3125612e403da73b4a1fe79c81c622b869df7cc4f97ce4ee910870229369b86a
  Stored in directory: /home/jovyan/.cache/pip/wheels/60/d9/c6/6e7909dfc39c636e7d25bec82a97d96cc82cd7d7462028e426
Successfully built uuid

[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: pip install --upgrade pip


In [1]:
from opensearchpy import OpenSearch

from datetime import datetime
import random
import string
import copy
from uuid import uuid4
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## What is OpenSearch?

OpenSearch is an **open-source, distributed search and analytics engine**. It enables developers, data scientists, and analysts to search, index, and analyze large volumes of data in near real-time. It's especially useful for full-text search, log analytics, and dashboarding.

- It supports full-text, fuzzy, and structured search.
- It scales horizontally across multiple machines.
- It integrates with Kibana's fork: **OpenSearch Dashboards**.
- It is open-source and community-driven, maintained by the OpenSearch Project (led by AWS).

You can learn more about the project at the official OpenSearch website: [https://opensearch.org](https://opensearch.org)

Here are a few great beginner-friendly videos to get you up to speed quickly:

1. **What is OpenSearch? (Official Intro by OpenSearch Project)**  
   https://www.youtube.com/watch?v=HOKBxcLHZ5A  
   A 3-minute official overview from the OpenSearch team.

2. **OpenSearch vs Elasticsearch: What's the Difference?**  
   https://www.youtube.com/watch?v=wNBvF0yUUvY  
   A helpful comparison for folks familiar with Elasticsearch.

3. **Search 101: How Search Engines Work (great conceptual foundation)**  
   https://www.youtube.com/watch?v=tj7lVdkS3Dg  
   This one is not OpenSearch-specific but gives great context on how search indexing and queries work in general.

OpenSearch is like a powerful search engine and analytics toolkit, built to help you quickly find and analyze data stored in JSON format.

In the next section, we’ll learn how to **install OpenSearch**, connect to it using Python, and send our first query!


## Trying OS on Local

We’ll use the official OpenSearch Docker images to run a single-node cluster locally.

This command pulls and runs OpenSearch in a Docker container with minimal configuration:

```bash
docker run -d --name opensearch \
  -p 9200:9200 -p 9600:9600 \
  -e "discovery.type=single-node" \
  -e "plugins.security.disabled=true" \
  opensearchproject/opensearch:2.13.0

## You will need to run this in your local to make sure you don't run into permission errors

```bash
!sudo usermod -aG docker $USER

### Docker SandBox

In [3]:
!docker run -d --name opensearch \
  -p 9200:9200 -p 9600:9600 \
  -e "discovery.type=single-node" \
  -e "plugins.security.disabled=true" \
  -e "OPENSEARCH_INITIAL_ADMIN_PASSWORD=StrongP@ssw0rd!" \
  opensearchproject/opensearch:2.13.0

docker: Error response from daemon: Conflict. The container name "/opensearch" is already in use by container "45fbdcc4afecd786860d74cc291412aca355c1b4fbaceb1c5d1624b48fabbfd6". You have to remove (or rename) that container to be able to reuse that name.

Run 'docker run --help' for more information


In [7]:
!docker ps

CONTAINER ID   IMAGE                                 COMMAND                  CREATED          STATUS          PORTS                                                                                                          NAMES
1904351271d0   opensearchproject/opensearch:2.13.0   "./opensearch-docker…"   24 seconds ago   Up 23 seconds   0.0.0.0:9200->9200/tcp, [::]:9200->9200/tcp, 9300/tcp, 0.0.0.0:9600->9600/tcp, [::]:9600->9600/tcp, 9650/tcp   opensearch


### Check to see if it's running

In [1]:
!curl -X GET "localhost:9200"

We **WON'T** be using the local version of OS for the rest of this demo, but I wanted to give it to you in case you want to explore a sandboxed version.

## Our Version

Giving you the password like this isn't great, but I have a feeling that no one has given you the contact information. "With great power, comes great....yada yada"

In [ ]:
# Pulling summarized docs from OpenSearch
from opensearchpy import OpenSearch, Urllib3HttpConnection, Urllib3AWSV4SignerAuth

client = OpenSearch(
    hosts = [{'host': ESHOST, 'port': 443}],
    http_auth = (ESUSER, ESPASSWORD),
    use_ssl = True,
    verify_certs = True,
)

client.info()

{'name': 'ea43be2afaab314f36308044c96406f7',
 'cluster_name': '181596953521:lmx-testing-2',
 'cluster_uuid': 'mivq_pOBR2GFTMJYdYELdg',
 'version': {'distribution': 'opensearch',
  'number': '2.17.0',
  'build_type': 'tar',
  'build_hash': 'unknown',
  'build_date': '2025-02-14T09:39:30.748828589Z',
  'build_snapshot': False,
  'lucene_version': '9.11.1',
  'minimum_wire_compatibility_version': '7.10.0',
  'minimum_index_compatibility_version': '7.0.0'},
 'tagline': 'The OpenSearch Project: https://opensearch.org/'}

## Understanding `client.info()` Output

When you call `client.info()` in OpenSearch, you're asking the server for basic metadata about itself — like version, identity, and configuration.

Here’s a breakdown of what each field means:

### Top-Level Fields

- **`name`**:  
  This is the name of the OpenSearch node that responded to your request.  
  In this case: `"50ce1a654b3c74459fce9d1ab68c8706"` — probably auto-generated.

- **`cluster_name`**:  
  The name of the entire OpenSearch cluster.  
  Here, it's `"181596953521:lmx-testing-2"` — often includes account or environment info.

- **`cluster_uuid`**:  
  A unique identifier for the cluster — useful for debugging or verifying consistency.


### `version` Block

This section tells you which version of OpenSearch is running, along with build details.

- **`distribution`**:  
  Confirms that it's an official `opensearch` distribution (not Elasticsearch or a custom fork).

- **`number`**:  
  This is the actual version of OpenSearch.  
  You're running version `2.17.0`.

- **`lucene_version`**:  
  OpenSearch is built on top of Apache Lucene, a low-level search library.  
  Lucene version here is `9.11.1`.

- **`build_type`, `build_hash`, `build_date`**:  
  Mostly useful for internal tracking or debugging (you can skip these for now).

- **`minimum_wire_compatibility_version` / `minimum_index_compatibility_version`**:  
  These show compatibility boundaries — for example, OpenSearch 2.17 can still talk to nodes as old as version 7.10.0.


### `tagline`

Just a nice reminder that you're using the OpenSearch project!  
`"The OpenSearch Project: https://opensearch.org/"`

### TL;DR

You’ve connected to a **live OpenSearch node** in a **cluster named `lmx-testing-2`**, running **version 2.17.0**. Everything looks healthy — now you're ready to create an index, add documents, and start querying.

## Indicies: How OS Manages Data


An **index** in OpenSearch is a logical namespace for storing and organizing documents. If you're familiar with relational databases, you can think of an index as roughly analogous to a table.

Each index contains a collection of **documents**, and each document is a structured JSON object.

Indices allow OpenSearch to organize, search, and analyze data efficiently.


### Why Indices Matter



OpenSearch uses indices to:

- Group related documents (e.g., logs, users, transactions)
- Apply specific settings for performance and scaling (shards, replicas, analyzers)
- Define mappings (similar to schemas) that declare how fields should be interpreted
- Apply lifecycle rules (e.g., how long data is retained or when it should be archived)


### Structure of an Index

Each index contains:

- **Documents**: JSON records of structured or semi-structured data.
- **Mappings**: The schema that defines the types of fields (e.g., text, keyword, date, integer).
- **Settings**: Shard and replica configurations, analyzers, and more.

You can create an index explicitly using the OpenSearch Python client. Here's an example that creates an index called `articles` with defined field types:

In [19]:
index_name = "demo"

def test_build():
    result = client.indices.create(
        index=index_name,
        body={
            "settings": {
                "number_of_shards": 2,
                "number_of_replicas": 2
            },
            "mappings": {
                "properties": {
                    "title": {"type": "text"},
                    "author": {"type": "keyword"},
                    "published": {"type": "date"},
                    "content": {"type": "text"}
                }
            }
        }
    )
    return result
    
test_build()

{'acknowledged': True, 'shards_acknowledged': True, 'index': 'demo'}

## Shards and Clusters in OpenSearch

### What is a Cluster?

An OpenSearch **cluster** is a group of one or more nodes (servers) that together hold your data and provide search and indexing capabilities.

- A single-node setup (like most local installations) is still considered a cluster.
- In a production system, multiple nodes work together for fault tolerance and scalability.

Each cluster has:
- A **unique name** (`cluster_name`)
- A **set of nodes**
- One **master node** (can be elected dynamically)
- Optional **data nodes**, **ingest nodes**, and **coordinating nodes**


### What is a Shard?

A **shard** is a horizontal partition of an index. Every index is split into one or more shards.

- Each shard is a fully functional and independent "mini-index."
- Shards allow OpenSearch to distribute data and load across multiple nodes.

There are two types of shards:

- **Primary shard**: Required for storing data. Every document belongs to exactly one primary shard.
- **Replica shard**: A copy of a primary shard. Provides redundancy and improves query throughput.

When you create an index, you can define:

Anyway, if you build an index that already exists, it will throw an error

In [20]:
try: 
    test_build()
except Exception as e:
    print(e)

RequestError(400, 'resource_already_exists_exception', 'index [demo/YmFSDmmJR82HwxvFqP6QEQ] already exists')


Here is how you delete an index....be careful with this one!

In [21]:
# Be careful with this one
# client.indices.delete(index=index_name, ignore=[400,404])

{'acknowledged': True}

## Find what we have now in OS

In [3]:
response = client.indices.get_alias(index="*")
[x for x in response.keys() if x[0] != "."]

['soft_string',
 'dev-model-description-index-test',
 'ai_string_tests',
 'edited_doc_36bu7cj4',
 'ai_string_tests_prod',
 'portfolio',
 'flattened_sheets',
 'dev-model-description-index']

This is what has already been made...I've been populating soft_string

In [4]:
mapping = client.indices.get_mapping(index="soft_string")
mapping

{'soft_string': {'mappings': {'dynamic_templates': [{'strings_as_text': {'match_mapping_type': 'string',
      'mapping': {'type': 'text'}}}],
   'properties': {'accuracy': {'type': 'long'},
    'actionability': {'type': 'long'},
    'ai_json': {'type': 'nested',
     'properties': {'Detailed Instructions': {'properties': {'Action': {'type': 'text'},
        'Field': {'type': 'text'},
        'Value': {'type': 'text'}}},
      'Details': {'properties': {'Common Issues': {'type': 'text'},
        'Field Instructions': {'properties': {'Field': {'type': 'text'},
          'Value': {'type': 'text'}}},
        'SAP T-code': {'type': 'text'},
        'User Permissions': {'type': 'text'}}},
      'Step Description': {'type': 'text'},
      'Step Expected Result': {'type': 'text'},
      'Step Number': {'type': 'long'}}},
    'ai_model': {'type': 'keyword'},
    'ai_scores': {'properties': {'accuracy': {'type': 'long'},
      'actionability': {'type': 'long'},
      'clarity': {'type': 'long'}

## Inspecting the `soft_string` Index Mapping

This document explains the field structure and indexing behavior of the `soft_string` index currently deployed in OpenSearch. The mapping reflects a legacy design with several planned improvements for future versions.

---

### Overview

The `soft_string` index schema includes:

- Nested fields for AI- and human-authored procedural steps
- Evaluation metrics generated by external models
- Metadata for document origin, authorship, and ingestion
- Standard fields for file paths, names, and other identifiers

This mapping is functional and expressive, but future iterations may optimize memory and performance by introducing stricter dynamic templates and replacing certain `text` fields with `keyword` where appropriate.

---

### Field Descriptions

#### 1. `ai_json` (type: `nested`)
- **Type:** `nested`
- **Purpose:** Stores structured AI-generated procedural steps.
- **Fields:**  
  - `Step Number`: long  
  - `Step Description`: text  
  - `Step Expected Result`: text  
  - `Details`:  
    - `SAP T-code`: text  
    - `Common Issues`: text  
    - `User Permissions`: text  
    - `Field Instructions`:  
      - `Field`: text  
      - `Value`: text  
  - `Detailed Instructions`:  
    - `Action`: text  
    - `Field`: text  
    - `Value`: text

#### 2. `ai_text` (type: `text` with `keyword` subfield)
- **Type:** text (analyzed)
- **Subfield:** keyword (non-analyzed, `ignore_above: 2048`)
- **Purpose:** Contains AI-generated commentary or critique.

#### 3. `ai_update` (type: `date`)
- **Type:** date
- **Format:** `yyyy-MM-dd`
- **Purpose:** Marks when AI-generated content was last updated.

#### 4. `clarity_sequence`, `accuracy`, `actionability`, `scores` (type: `long`)
- **Type:** long
- **Purpose:** Evaluation scores generated by an external model for:
  - Step clarity and sequencing
  - Factual accuracy
  - Practical utility
  - Overall summary score

#### 5. `created`, `modified` (type: `date`)
- **Format:** `yyyy-MM-dd`
- **Purpose:** Timestamps for creation and last modification.

#### 6. `creator`, `last_modified_by` (type: `text`)
- **Type:** text
- **Purpose:** Indicates who created or edited the document.
- **Note:** These use `text` rather than `keyword` due to legacy design.

#### 7. `extension`, `name`, `title`, `sheet_name`, `information`, `keywords`, `other`, `path`, `path_name` (type: `text`)
- **Purpose:** Capture file metadata and categorization terms.
- **Note:** Only `title` includes a `keyword` subfield. Others may be changed to `keyword` in future versions.

#### 8. `step_json` (type: `nested`)
- **Type:** nested
- **Purpose:** Stores human-authored procedural steps.
- **Structure:** Mirrors `ai_json`, with `Step Number`, `Step Description`, `Step Expected Result`, `Details`, and `Detailed Instructions`.

#### 9. `metrics` (type: `nested`)
- **Type:** nested
- **Purpose:** Captures time-series metrics or auxiliary analysis results.
- **Fields:**  
  - `name`: text  
  - `value`: float  
  - `note`: text  
  - `timestamp`: date

#### 10. `status`, `review_status`, `reviewed_by`, `tags`, `schema_version`, `source_pipeline`, `ai_model` (type: `keyword`)
- **Type:** keyword
- **Purpose:** Categorical metadata to track model version, review status, processing pipeline, etc.

#### 11. `step_count`, `token_count`, `first_sheet` (type: `integer` or `long`)
- **Purpose:** Internal counters for number of steps, token length, or sheet positioning.

#### 12. `ingest_time` (type: `date`)
- **Type:** date
- **Format:** Expected to follow `yyyy-MM-dd`, though OpenSearch default parsing supports higher precision.
- **Purpose:** Tracks when the document was added to the index.

---

### Dynamic Template

```json
{
  "dynamic_templates": [
    {
      "strings_as_text": {
        "match_mapping_type": "string",
        "mapping": { "type": "text" }
      }
    }
  ]
}```

- **Purpose:** Ensures that all fields with type `string` are automatically mapped as `text`.
- **Behavior:** This allows full-text search across arbitrary fields not explicitly defined in the mapping.
- **Note:** This is a legacy choice. In future revisions, a more constrained or explicit mapping approach is expected to reduce memory usage and improve indexing performance.

---

### Summary

The `soft_string` index schema supports a comprehensive pipeline for AI-enriched document processing and metadata capture. Key features include:

- **Structured AI and human steps** using `nested` mappings in `ai_json` and `step_json`
- **Model-driven evaluation metrics** such as `clarity_sequence`, `accuracy`, and `actionability`
- **Comprehensive metadata** for file tracking (`path`, `name`, `title`, `extension`, etc.)
- **Version and review tracking** through `status`, `tags`, `review_status`, and `source_pipeline`
- **Metric support** for numeric trends and observations through the `metrics` nested field
- **Temporal tracking** using standardized `date` fields like `created`, `modified`, and `ingest_time`

This mapping is designed for flexibility and is currently in use, but a future version may include:

- A shift toward `keyword` fields for identifiers and categoricals
- Reduced dynamic field generation
- Optimization of nested structures to minimize query cost

Understanding this structure allows for precise querying, effective enrichment, and extensible data management within OpenSearch.

Honestly, I am not sure if this a perfect mapping and once you get familar with everything, we can tag-team how to make this better. But for now, it's what we got.

## Sample on the Index

In [25]:
query = {
    "query": {
        "match_all": {}
    },
    "size": 20, # Number of documents to return in the sample
}

results = client.search(index="soft_string", body=query)

In [26]:
print(results.keys())
print(results["hits"].keys())
print(results["hits"]["hits"][0].keys())

dict_keys(['took', 'timed_out', '_shards', 'hits'])
dict_keys(['total', 'max_score', 'hits'])
dict_keys(['_index', '_id', '_score', '_source'])


## Interpreting the OpenSearch Search Response

When you execute a search query in OpenSearch using the Python client, the response is a dictionary that contains multiple layers of metadata and actual results. Here’s how to interpret it:

### Top-Level Fields

- **`took`**:  
  Indicates how many milliseconds the query took to execute on the server.  
  This can be used to measure search performance.

- **`timed_out`**:  
  Boolean flag — `False` means the query completed successfully within the timeout window.

- **`_shards`**:  
  Details about how the query was distributed across the shards.
  - `total`: The number of shards queried (based on index configuration).
  - `successful`: How many shards returned results successfully.
  - `failed`: Number of shards that failed to respond (ideally 0).
  - `skipped`: Number of shards skipped due to routing or filtering.

---

### `hits` Object

This section contains the actual search results.

- **`total`**:  
  The total number of documents that matched the query.
  - `value`: The number of matched documents (e.g., 653).
  - `relation`: Can be `eq` (exact) or `gte` (estimated lower bound).

- **`max_score`**:  
  The highest `_score` of all returned hits. Useful for relevance ranking.

- **`hits`**:  
  A list of documents (usually truncated to the top `size` results).

---

### Each Hit in the `hits` List

Each item in the list is a matching document with additional metadata:

- **`_index`**:  
  The name of the index the document belongs to (`ai_string_tests` in this case).

- **`_id`**:  
  The unique document ID assigned by OpenSearch.

- **`_score`**:  
  A relevance score (used mostly in `match` or `multi_match` queries).  
  A score of 1.0 is common in `match_all` or `term` queries with no ranking.

- **`_source`**:  
  The actual content of the document as it was indexed. This includes all the fields such as:
  - `step_json` and `ai_json`: Nested JSON containing structured process steps.
  - `created`, `modified`: Dates of creation and last modification.
  - `name`, `path`, `sheet_name`: File-related metadata.
  - `ai_text`: A large blob of text with critique and reformatted JSON.
  - `title`, `creator`, `keywords`: Metadata for sorting, filtering, or searching.

---

### Summary

This structure allows OpenSearch to return both metadata (for performance monitoring and routing) and actual content (for your application logic or frontend display).

You can access just the documents using:

In [27]:
results = client.search(index="soft_string", body={"query": {"match_all": {}}, "size": 20})
docs = [hit["_source"] for hit in results["hits"]["hits"]]

In [28]:
# This is the 2nd item on the list via python list indexing
docs[1]

{'step_json': [{'Step Number': 1,
   'Step Description': 'Log into the S4 environment being used for this test cycle and run the program to load the projects and wbs structures to be converted  (PLP-90069-C)',
   'Step Expected Result': 'Projects and WBS Structure were loaded successfully; capture the number of WBS Elements that are in the file to load and the number that were successfully loaded; capture any errors to be reported to the data validator so that any issues can be corrected prior to the next load.  If desired, issues can be corrected within the current file and reloaded to improve the load percentage.',
   'Details': {'SAP T-code': '',
    'Field Instructions': {'Field': '', 'Value': ''}},
   'Detailed Instructions': {'Action': '', 'Field': '', 'Value': ''}},
  {'Step Number': 2,
   'Step Description': 'Log into the S4 environment being used for this test cycle',
   'Step Expected Result': 'Log in successful',
   'Details': {'SAP T-code': '',
    'Field Instructions': {'F

## Exact Search on Path String

The `match_phrase` query looks for the *exact sequence of terms* in the specified field.
It is different from a regular `match`, which allows term reordering and partial matches.
This query will return documents where the `path` field contains the exact phrase "itc-2",
with terms in that specific order.

In [31]:
query = {
  "query": {
    "match_phrase": {
      "path": "itc-2"
    }
  }, "size": 10_000
}
results = client.search(index="soft_string", body=query)

In [36]:
results = client.search(index="soft_string", body=query)
print("Random 10 Documents with exact phrase 'itc-2' in the path field:")
random.shuffle(results["hits"]["hits"])
for num, hit in enumerate(results["hits"]["hits"][:10]):
    print(f'{num:01}\t{hit["_source"]["path"][0:75]:5}...')

Random 10 Documents with exact phrase 'itc-2' in the path field:
0	Data/3.  Integration Test Repository/ITC-2/Finance/DELETE/INT_RTR_005A_DEL/...
1	Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_001 ...
2	Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_006 ...
3	Data/3.  Integration Test Repository/ITC-2/Finance/INT_RTR_005A/01.TC_O2C_I...
4	Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_005 ...
5	Data/3.  Integration Test Repository/ITC-2/Finance/INT_RTR_005B/01.TC_O2C_I...
6	Data/3.  Integration Test Repository/ITC-2/Finance/DELETE/INT_RTR_005A_DEL/...
7	Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_002 ...
8	Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_002 ...
9	Data/3.  Integration Test Repository/ITC-2/Finance/DELETE/INT_RTR_005A_DEL/...


### Dot notation example for querying inside a nested object


If your mapping contains a structured JSON object like `step_json` or `ai_json`,
you can access subfields using dot notation.
For example: step_json.0.Step Description
Note: This assumes step_json is NOT mapped as a nested type. If it is, you must use a nested query.

### Fuzzy search inside a JSON string field using dot notation example
and an exact phrase match on the 'path' field.

In [69]:
# This query performs:
# - A fuzzy match inside a nested field (ai_json)
# - An exact match_phrase on the 'path' field

nested_fuzzy_query = {
  "query": {
    "bool": {
      "must": [
        {
          "match_phrase": {
            "path": "itc-2"
          }
        },
        {
          "nested": {
            "path": "step_json",
            "query": {
              "bool": {
                "must": [
                  {
                    "match_phrase": {
                      "step_json.Step Description": "Validate"
                    }
                  },
                  {
                    "match_phrase": {
                      "step_json.Step Expected Result": "All fields are populated correctly"
                    }
                  }
                ]
              }
            },
            "inner_hits": {
              "size": 5
            }
          }
        }
      ]
    }
  },
  "_source": ["name", "path", "step_json"]
}

response = client.search(index="soft_string", body=nested_fuzzy_query)

print("Results for fuzzy match inside ai_json (nested) and exact phrase 'itc-2' in path:")
for hit in response["hits"]["hits"]:
    print("Path:", hit["_source"].get("path"))
    print("Step 0 Description:", hit["_source"].get("ai_json", {}).get("0", {}).get("Step Description", ""))
    print("---")

Results for fuzzy match inside ai_json (nested) and exact phrase 'itc-2' in path:
Path: Data/3.  Integration Test Repository/ITC-2/Prod Ops/INT_PTS_003 - Execute Manufacturing - Build - External Procurement/Test Cases/00.TC_O2C_Initiate and Monitor Contract_001 - PTS.xlsx
Step 0 Description: 
---
Path: Data/3.  Integration Test Repository/ITC-2/Contracts & Billing/INT_O2C_005 - Customer Contract Creation through Customer Billing and Apply Cash (Deliverable Material, Fixed Price, Progress Payment)/01.TC_O2C_Initiate and Monitor Contract_001 - O2C.xlsx
Step 0 Description: 
---
Path: Data/3.  Integration Test Repository/ITC-2/Prod Ops/Archive/INT_PTS_003 - Execute Manufacturing - Build - External Procurement/Test Cases/00. STR_CAB_CLM_Initiate and Monitor Contract.xlsx
Step 0 Description: 
---
Path: Data/3.  Integration Test Repository/ITC-2/Finance/INT_RTR_008B/01.TC_O2C_Initiate and Monitor Contract_01.xlsx
Step 0 Description: 
---
Path: Data/3.  Integration Test Repository/ITC-2/Contra

## Explanation of Nested and Exact Match Query Using Dot Notation

This query demonstrates a combined search strategy in OpenSearch that targets both top-level fields and nested fields using dot notation. The goal is to retrieve documents that match criteria in the main document structure and within a nested array of objects.

The query uses a `bool` structure with multiple `must` clauses to enforce that all specified conditions must be true for a document to match.

First, the `path` field is searched using a phrase-based query. This field exists at the top level of the document and contains structured path data. The phrase match allows us to identify documents where the value includes the term "itc-2", which is treated as a meaningful, possibly case-insensitive phrase within the indexed text. Although not an exact keyword match, this type of search is sufficient for matching specific terms within analyzed `text` fields.

Next, a `nested` query is used to examine individual entries in the `step_json` array. This field is explicitly mapped as `nested`, meaning OpenSearch treats each element in the array as its own subdocument. This allows us to construct queries that apply to each step individually, rather than the array as a whole.

Inside the `nested` query, two conditions are applied. One searches within the `Step Description` field for the presence of the word "Validate" using a phrase match. The second searches within the `Step Expected Result` field for the exact phrase "All fields are populated correctly". These conditions are written using dot notation, which allows precise access to deeply nested fields, such as `step_json.Step Description` and `step_json.Step Expected Result`.

The use of dot notation here is critical. It tells OpenSearch to look specifically within the nested path defined by the `step_json` field. Without this, the query would not correctly target the subfields within the nested structure, and matches would not be evaluated per-object.

The query also includes an `inner_hits` clause, which returns the specific nested objects that matched. This is especially useful for verifying which steps within the array satisfied the search conditions, rather than only seeing that a document matched overall.

Lastly, a `_source` filter is applied to limit the returned fields. This optimizes response size by retrieving only the document name, path, and step content relevant to the query.

In summary, this query combines top-level filtering and nested field searching in a unified structure, using dot notation to precisely navigate and query fields deep within the document schema.

## = WORKFLOW: USER-DRIVEN JSON EDITING IN OPENSEARCH =

This script models the process your app will perform:
1. Search for a document by exact match (e.g., title)
2. Extract and copy the document for editing
3. Modify the step_json field (simulate user edits)
4. Add a note for auditing (e.g., edit timestamp)
5. Insert the edited document into a new index
I don't think this is a perfect mirror of what we are hoping to do, but we can talk through it. Or you can 

### Step 1: Perform an exact match search on the path field


In [124]:
search_query = {
    "query": {
        "match_phrase": {
            "path": "Integration Test Repository/ITC-2/Prod Ops/Archive/INT_PTS_003 - Execute Manufacturing - Build - External Procurement/Test Cases/00. STR_CAB_CLM_Initiate and Monitor Contract.xlsx" }
    }
}

search_result = client.search(index="soft_string", body=search_query)

In [125]:
search_result["hits"]['total']

{'value': 1, 'relation': 'eq'}

### Step 2 and 3: Extract and Modify matching document

In [108]:
if not search_result["hits"]["hits"]:
    print("No document found with that title.")
else:
    original_doc = search_result["hits"]["hits"][0]["_source"]
    print("Original document retrieved.")

    # Step 3: Deep copy and modify the JSON
    original_id = search_result["hits"]["hits"][0]["_id"]
    modified_doc = copy.deepcopy(original_doc)
    
# Simulate user editing step_json (you can replace this with UI-driven input later)
if "step_json" in modified_doc and isinstance(modified_doc["step_json"], list):
    modified_doc["step_json"].append({
        "Step Number": 23,
        "Step Description": "User-inserted step from editor",
        "Step Expected Result": "Custom logic executed",
        "Details": {
            "SAP T-code": "",
            "Field Instructions": {
                "Field": "",
                "Value": ""
            }
        },
        "Detailed Instructions": {
            "Action": "",
            "Field": "",
            "Value": ""
        }
    })
    
    # For this example, I am also going to add some fake metrics information to
    
    modified_doc["metrics"] = [{"name": "score_test", 
                                "value": 9001, "note": 
                                "What over 9000, there is no way that can be right",
                                "timestamp": datetime.now().timestamp() * 1000}]
    
    modified_doc["ai_update"] = datetime.now().strftime('%Y-%m-%d')
    modified_doc["accuracy"] = 2

Original document retrieved.


In [109]:
# Option A: Overwrite existing document by reusing its _id
overwrite = False  # Set to True if you want to replace the original

if overwrite:
    response = client.index(index="soft_string", id=original_id, body=modified_doc)
    print(f"Document overwritten (same ID): {original_id}")
else:
    new_id = str(uuid4())  # Generate a unique ID
    response = client.index(index="soft_string", id=new_id, body=modified_doc)
    print(f"Modified document inserted with new ID: {new_id}")

Modified document inserted with new ID: 83d72271-3c77-4a4f-acda-3cfbd91b41a8


## Review to see if we made the change

This is how you search by index

In [116]:
response = client.get(index="soft_string", id=new_id)

In [117]:
response["_source"]["step_json"][-1]

{'Step Number': 23,
 'Step Description': 'User-inserted step from editor',
 'Step Expected Result': 'Custom logic executed',
 'Details': {'SAP T-code': '',
  'Field Instructions': {'Field': '', 'Value': ''}},
 'Detailed Instructions': {'Action': '', 'Field': '', 'Value': ''}}

In [118]:
response["_source"]["metrics"]

[{'name': 'score_test',
  'value': 9001,
  'note': 'What over 9000, there is no way that can be right',
  'timestamp': 1743685755465.6392}]

In [119]:
response["_source"]["accuracy"]

2

## Delete out Example by _ID

In [120]:
## Be careful with this one
response = client.delete(index="soft_string", id=new_id)
print(response)

{'_index': 'soft_string', '_id': '83d72271-3c77-4a4f-acda-3cfbd91b41a8', '_version': 2, 'result': 'deleted', '_shards': {'total': 3, 'successful': 3, 'failed': 0}, '_seq_no': 6, '_primary_term': 1}


In [121]:
try:
    response = client.get(index="soft_string", id=new_id)
except Exception as e:
    print(e)

NotFoundError(404, '{"_index":"soft_string","_id":"83d72271-3c77-4a4f-acda-3cfbd91b41a8","found":false}')
